### ISSUE TITLE
condition cache in wrong state when factory is shared between workflows 

### ISSUE THREAD

**Issue by oliwel** [jonasbn/perl-workflow#9]
The condition cache is part of the global factory. If you have more than one workflow of same type from this factory - either by running them in parallel or by handling them one after another using a persister, the workflow cache is NOT cleared prior to "execute_action".

The best fix IMHO would be an explicit call to clear_condition_cache as first command in Workflow::execute_action.

I also think that the "clear_condition_cache" in 133ff should ALWAYS be executed if the transaction is not rolledback - it is a legal setup to have an action that points back to the original state. In this case the cache wont be cleared. But: If I am right, this call can be removed when we place it instead in the beginning of execute_action.

**Comment by ehuelsmann** [comment]
@jonasbn , you added `hacktoberfest`. While it's not October, I'm willing to take this on. However, I would definitely benefit from some direction as to the solution. Should the condition cache move to the workflow itself instead of being maintained in the factory?

**Comment by ehuelsmann** [comment]
Isn't this a duplicate of #7 ?

**Comment by ehuelsmann** [comment]
And a duplicate of #5 ?

**Comment by ehuelsmann** [comment]
@oliwel, when going over the intention of the condition cache, I don't think clearing the condition cache at the beginning of `evaluate_action()` is the solution, because it is used to determine `is_action_available()`; the condition cache is meant to stabilize the outcome of a series of queries against `is_action_available()`, as long as the workflow is in a specific state.

I'm going over the code base and the origins of the condition cache now to see if there's a better solution.

**Comment by oliwel** [comment]
@ehuelsmann #7 is exactly the same problem, #5 is just a flag to turn off the caching, I dont see a benefit on that besides healing this bug ;)

Moving this to the factory does not help, we fixed this in our project by adding a clear_condition_cache call in Workflow->init and in Workflow::Action->execute_action which helped while keeping the cache intact at least as far as we observed and tested this. I am happy to move the fixes to upstream and make a pull request - I havent done this yet as I was not sure if this has side effects but as it works for more than 4 years now in our product I'd say yes.

**Comment by jonasbn** [comment]
Hi @oliwel 

Please, please, please send a patch upstream, I would love for this issue.

With the recent contributions and initiative by @ehuelsmann, I think we need to get the distribution clean, polished and stabilized.

**Comment by ehuelsmann** [comment]
@jonasbn I've been studying the code base over the past days. I was wondering if it isn't better to move the cache from the Workflow::State object (which is somewhat global) into the Workflow instance, which has a much more restricted scope for caching. I can submit code to that effect, maybe already tomorrow.

Another thought that I was having is that the current implementation (of the cache) has some unfortunate side effects, but I'm still working on my understanding so I'll be able to articulate my problem.

**Comment by oliwel** [comment]
@ehuelsmann  (cc @jonasbn ) it might be benefical to have a realtime chat on this - we messed a lot around with the cache functionality and hacked some internals so I ultimately  want to avoid to break our codebase and likely invent other issues..

**Comment by jonasbn** [comment]
@oliwel and @ehuelsmann 

I have a Slack available if need be, else IRC is my preferred choice. Gitter could also be an option, went for Slack due to the nifty integrations though.

**Comment by oliwel** [comment]
Its some time ago that I fired up IRC - is there really more than background noise going on :) 
We started to use keybase.io in our (keybase.io/inv/b356e0ae60) project but I am also happy to use Slack or IRC. We just need to arrange a timeslot... any time tomorrow between 14 and 22 CET would fit for me.

**Comment by ehuelsmann** [comment]
I can use Slack, keybase.io and IRC. I think that Slack seems to be the common denominator. Tomorrow between 14 and 22 CET suits me well too, although I may have to tend to other obligations inbetween.

**Comment by jonasbn** [comment]
I do not know keybase, I will check it out at some point, I have mailed you both invites to the Slack for this repository

**Comment by ehuelsmann** [comment]
PR #27 implements the condition cache as I envisioned it when I brought up the topic yesterday in reaction to @oliwel's proposal to add a few more cache clearing points.

PR #28 is my reaction to @oliwel's statement that the cache would need no more locality than `get_current_actions()` or `execute_action()`. My personal preference is this PR because it eliminates the need to manage the cache through specifically triggered "clearing points".

**Comment by jonasbn** [comment]
I prefer the concept in PR #27 over the one in #28
